### DISCUSSION TITLE
JVM의 메모리 영역에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#82] [Java]
어떤 종류가 있는지, 각각의 특성

**Comment by le2sky** [comment]
https://github.com/2024-woowacourse-study/level-interview/discussions/39

이것과 연관 지을수도 있겠네요. 그리고 아래도 같이 공부해보면 좋을듯?

+ JIT 컴파일 ?
+ GC + JVM 힙 메모리
+ String Pool

**Comment by J-I-H-O** [comment]
런타임 데이터 영역은 아래 종류가 있습니다.

**메서드 영역(Method Area)**
각 클래스의 구조, 필드 및 메서드에 대한 정보를 저장하는 영역입니다.
JVM이 시작될 때 생성되는 공간으로, 프로그램 종료 시까지 유지됩니다. 모든 쓰레드가 공유한다는 특징이 있습니다.

**힙(Heap)**
동적으로 할당되는 메모리를 저장하는 영역입니다. new 키워드로 생성되는 객체 인스턴스 및 배열 등을 저장합니다.
메서드 영역과 함께 모든 쓰레드가 공유하며, 런타임 시 동적으로 할당하여 사용됩니다.

**스택(Stack)**
Primitive Type을 저장하는 공간으로, 임시적으로 사용되는 변수나 정보들이 저장되는 영역입니다. Primitive Type 변수는 스택 영역에 직접 값을 가지지만, 참조타입 변수는 힙 영역이나 메서드 영역의 참조 주소를 갖습니다.
스택 영역은 각 쓰레드마다 하나씩 존재하며, 쓰레드가 시작될 때 할당됩니다.

**PC레지스터(Program Counter Register)**
PC레지스터는 현재 수행중인 JVM 명령어 주소를 가리키는 포인입니다. 쓰레드가 시작될 때 생성되고, 쓰레드마다 별도로 유지됩니다.

**네이티브 메서드 스택(Native method Stack)**
기계어로 작성된 프로그램을 실행시키는 영역입니다. 또, 자바 이외의 언어로 작성된 네이티브 코드를 실행하기 위한 공간이기도 합니다.

**Comment by 3Juhwan** [comment]
JVM의 메모리 영역은 Java 프로그램이 실행될 때 사용되는 메모리를 관리하는 데 중요한 역할을 합니다. 주요한 메모리 영역은 다음과 같습니다.

힙: 객체가 저장되는 공간으로, 모든 Java 객체는 힙에 할당됩니다. 힙은 Young Generation, Old Generation, 그리고 Perm Generation으로 나뉩니다. Young Generation에는 새로 생성된 객체들이 할당되고, Old Generation에는 오랫동안 살아남은 객체들이 할당됩니다. Perm Generation에는 클래스와 메타데이터 등이 저장됩니다. Java 8부터는 Perm Generation이 Metaspace로 대체되었습니다.

스택: 메소드 호출과 관련된 정보를 저장하는 공간으로, 메소드가 호출될 때마다 스택 프레임이 생성되고 메소드의 지역 변수, 매개 변수, 복귀 주소 등이 저장됩니다. 스택은 스레드마다 별도로 존재하며, 메소드 호출이 완료되면 해당 스택 프레임이 제거됩니다.

메소드 영역: 클래스 파일의 바이트 코드, 메소드와 필드 정보, 정적 변수 등이 저장되는 공간입니다. 모든 스레드가 공유하며 JVM이 시작될 때 생성되며, 프로그램이 종료될 때까지 유지됩니다. Java 8부터는 Perm Generation이 Metaspace로 대체되었습니다.

PC 레지스터: 현재 수행 중인 JVM 명령의 주소를 저장하는 공간입니다. 각 스레드마다 별도의 PC 레지스터가 있으며, 스레드가 실행될 때마다 생성되고 종료될 때 제거됩니다.

네이티브 메소드 스택: 네이티브 코드를 실행하기 위한 스택으로, 네이티브 메소드 호출에 필요한 정보를 저장합니다.

**Comment by GIVEN53** [comment]
힙 - 객체 저장, YG, OG, Metaspace
메서드 - 클래스/메서드 정보, static 등 저장, 종료까지 유지
스택 - 원시타입, 지역 번수 등 저장, 메서드 호출 시 스택 프레임, 스레드마다 별개
PC 레지스터 - 명령어 주소 저장, 스레드마다 별개
네이티브 메서드 스택 - 네이티브 코드를 실행하기 위한 스택
